lib/pos_mempool.go

```diff
@@ -0,0 +1,490 @@
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	if !dmp.IsRunning() {
+		return
+	}
+
+	// Close the persister and stop the database.
+	if err := dmp.persister.Stop(); err != nil {
+		glog.Errorf("PosMempool.Stop: Problem stopping persister: %v", err)
+	}
+	if err := dmp.db.Close(); err != nil {
+		glog.Errorf("PosMempool.Stop: Problem closing database: %v", err)
+	}
+	// Reset the transaction register, the ledger, and the nonce tracker.
+	dmp.txnRegister.Reset()
+	dmp.ledger.Reset()
+	dmp.nonceTracker.Reset()
+
+	dmp.status = PosMempoolStatusNotRunning
+}
+
+func (dmp *PosMempool) IsRunning() bool {
+	return dmp.status == PosMempoolStatusRunning
+}
+
+// AddTransaction validates a MsgDeSoTxn transaction and adds it to the mempool if it is valid.
+// If the mempool overflows as a result of adding the transaction, the mempool is pruned.
+func (dmp *PosMempool) AddTransaction(txn *MsgDeSoTxn) error {
+	// First, validate that the transaction is properly formatted according to BalanceModel. We acquire a read lock on
+	// the mempool. This allows multiple goroutines to safely perform transaction validation concurrently. In particular,
+	// transaction signature verification can be parallelized.
+	if err := dmp.validateTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem verifying transaction")
+	}
+
+	// If we get this far, it means that the transaction is valid. We can now add it to the mempool.
+	// We lock the mempool to ensure that no other thread is modifying it while we add the transaction.
+	dmp.Lock()
+	defer dmp.Unlock()
+
+	if !dmp.IsRunning() {
+		return errors.Wrapf(MempoolErrorNotRunning, "PosMempool.AddTransaction: ")
+	}
+
+	// Construct the MempoolTx from the MsgDeSoTxn.
+	mempoolTx, err := NewMempoolTx(txn, dmp.latestBlockHeight)
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem constructing MempoolTx")
+	}
+
+	// Check the nonceTracker to see if this transaction is meant to replace an existing one.
+	pk := NewPublicKey(txn.PublicKey)
+	if existingTxn := dmp.nonceTracker.GetTxnByPublicKeyNonce(*pk, *txn.TxnNonce); existingTxn != nil {
+		if existingTxn.FeePerKB > mempoolTx.FeePerKB {
+			return errors.Wrapf(MempoolFailedReplaceByHigherFee, "PosMempool.AddTransaction: Problem replacing transaction "+
+				"by higher fee failed. New transaction has lower fee.")
+		}
+		if err := dmp.removeTransactionNoLock(existingTxn, true); err != nil {
+			return errors.Wrapf(err, "PosMempool.AddTransaction: Problem removing old transaction from mempool during "+
+				"replacement with higher fee.")
+		}
+	}
+
+	// Add the transaction to the mempool and then prune if needed.
+	if err := dmp.addTransactionNoLock(mempoolTx, true); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem adding transaction to mempool")
+	}
+
+	// Add the transaction to the nonce tracker.
+	dmp.nonceTracker.AddTxnByPublicKeyNonce(*pk, *txn.TxnNonce, mempoolTx)
```

Comment:
Why isn't the nonceTracker update just a part of removeTransactionNoLock() and addTransactionNoLock() like the other things? Seems like you could do the following:
```
1. Keep the check above where you remove an existing txn if it has a matching nonce. This will remove it from nonceTracker.

2. Add the new txn with addTransactionNoLock, which will add the new txn's nonce to the nonceTracker.
```